File: lib/glg/SimpleViewer/SingleThreaded/C/demo_datafeed.c
#include "GlgViewer.h"

extern GlgObject Viewport;

/*--------------------------------------------------------------------
| Generate simulated D value in the range defined by Low and High variables.
| If RANDOM_DATA is False, ReadDValue from datafeed.c will be used
| to obtain real-time value.
*/
GlgBoolean DemoReadDValue( TagRecord * tag_record, double * d_value )
{

   /* Data range for simulated demo data. */
   double low = 0.;
   double high = 100.;
   
   double period = 100;
   double
     alpha, center, half_amplitude;

   static long counter = 0;

   period = 500.;
   half_amplitude = ( high - low ) / 2.0;
   center = low + half_amplitude;
   alpha = 2. * M_PI * counter / period;
   *d_value = center + half_amplitude * sin( alpha ) * sin( alpha / 30. );

   ++counter;
   return GlgTrue;
}

/*--------------------------------------------------------------------
| Generate simulated S value
*/
GlgBoolean DemoReadSValue( TagRecord * tag_record, char ** s_value )
{
   if( !s_value )
     return GlgFalse;
   
   *s_value = GlgRand( 0., 1. ) > 0.5 ? "On" : "Off";

   return GlgTrue;
}

/*--------------------------------------------------------------------
| Simulte the "Write" opearion. For demo purposes, set the value
| of the specified tag source in the drawing. 
*/
GlgBoolean DemoWriteDValue( char * tag_source, double d_value )
{
   return GlgSetDTag( Viewport, tag_source, d_value, False );  
}

File: lib/glg/SimpleViewer/SingleThreaded/Java/DemoDataFeed.java
import com.genlogic.*;

//////////////////////////////////////////////////////////////////////////
// DemoDataFeed provides simulated data for demo, as well as for testing 
// with no LiveDataFeed.
// In an application, data will be coming from LiveDataFeed.
//////////////////////////////////////////////////////////////////////////

public class DemoDataFeed implements DataFeedInterface
{
   GlgViewer Viewer;

   long counter = 0;
   long tag_index = 0;
   double High = 100.;
   double Low = 0;

   public DemoDataFeed( GlgViewer viewer ) 
   {
      Viewer = viewer;
   }

   /////////////////////////////////////////////////////////////////////// 
   // Generate demo data value of type D (double)
   ///////////////////////////////////////////////////////////////////////
   public boolean ReadDValue( GlgTagRecord tag_record, DataPoint data_point )
   {
      double 
        half_amplitude, center,
        period = 100.0,
        alpha;
      
      half_amplitude = ( High - Low ) / 2.0;
      center = Low + half_amplitude;
      
      alpha = 2.0 * Math.PI * counter / period;
      
      data_point.d_value = center + 
        half_amplitude * Math.sin( alpha ) * Math.sin( alpha / 30.0 );
      
      ++counter;
      return true;
   }

   /////////////////////////////////////////////////////////////////////// 
   public boolean ReadSValue( GlgTagRecord tag_record, DataPoint data_point )
   {
      data_point.s_value = "";
      return true;
   }

   /////////////////////////////////////////////////////////////////////// 
   public boolean WriteDValue( String tag_source, double d_value )
   {
      /* In the demo mode, set the tag in the drawing to simulate round trip
         after the write operation.
      */
      return Viewer.SetDTag( tag_source, d_value, false );
   }

   /////////////////////////////////////////////////////////////////////// 
   public boolean WriteSValue( String tag_source, String s_value )
   {
      /* In the demo mode, set the tag in the drawing to simulate round trip
         after the write operation.
      */
      return Viewer.SetSTag( tag_source, s_value, false );
   }
}
